DAOS-13564 object: add obj flatten APIs

liuxuezhao opened the pull request:
1. add daos_cont_set_ro() API definition
2. add DAOS_PROP_CO_READONLY container status bit
3. rename daos_cont_status_clear() to daos_cont_set_healthy()
    add daos_cont_status_clear alias to keep back-compatibility
4. define DAOS_PROP_CSF_HEALTHY/DAOS_PROP_CSF_READONLY internal flag bit
    cleanup DAOS_PROP_CO_CLEAR as no usage in code.
5. add daos_obj_set_ro() API definition

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

liuxuezhao committed:
DAOS-13564 object: add obj flatten APIs

1) DAOS_IOF_FLATTEN flag carried daos_obj_update()
2) daos_obj_flatten()

Required-githooks: true

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liuxuezhao commented on src/include/daos_obj.h:
```diff
@@ -781,7 +798,7 @@ daos_obj_fetch(daos_handle_t oh, daos_handle_t th, uint64_t flags,
  *			Use DAOS_TX_NONE for an independent transaction.
  *
- * \param[in]	flags	Update flags (conditional ops).
+ * \param[in]	flags	Update flags (conditional ops or DAOS_IOF_FLATTEN).
```
add DAOS_IOF_FLATTEN flag for daos_obj_update() because in wiki page it says -
"DAOS can provide API to allow user to indicate immutability of WORM objects:
   Application/middleware calls write(obj) with “DONE” flag."

But adding this flag to normal update request will make the obj_update() be more complex - for example request targets are different (normal update only to one dkey/group, but flatten to all groups' shards) this will make some detailed code not very convenient to handle, and possibly cause update RPC interoperability issue.

How about remove this flag's usage, only provide daos_obj_flatten() API?

mchaarawi replied:
yes this would even be worse coming in from the dfs or array api where an "update" can turn into multiple daos_obj_update(). so just a single API call is better here.
that being said, the name flatten is ambiguous and has several meanings from a user perspective. there is no indication here that flatten means make immutable.
so i would suggest to rename flatten to some more informative for the user like daos_obj_set_immutable() or set_ro(). that being said, what if we want to add more properties later on the object?
maybe better to just add a new API:
daos_obj_set_prop(oid, flags);
and DAOS_OBJ_RO being one of the flags (only one for now)?

liuxuezhao replied:
Ok, let's only add a single API if no objection from others.

I think probably no plan to add object level property for now.
So maybe just one API - daos_obj_set_ro () is good...

mchaarawi replied:
sure, until there becomes a plan to add more ;-)
so IMO better to plan ahead always for APIs rather than make API changes later.

NiuYawei replied:
I agree that single API like daos_obj_set_prop(DAOS_OBJ_PROP_RO) looks better.

liuxuezhao replied:
daos_obj_set_prop() looks will mislead user think daos will provide object level property, I think that is not true because object level property will cause too much metadata , not we only have property for pool and container level.

mchaarawi replied:
i see. anyway no strong opinion on this one. both ways sound OK to me.
i'd like to discuss though how this applies in a more generic sense, like "flattening" an entire container. we cannot expect to drive this all from the client side per object. i think better to discuss this on a meeting for now.

liuxuezhao replied:
in last week's md on ssd meeting, @gnailzenh suggested to start with container level flattening (flaten all objects for the container).

liuxuezhao replied:
@mchaarawi added container property related change for it, please check if it is fine wit you, thanks.

liuxuezhao committed:
DAOS-13564 object: add daos_obj_set_ro() API

Required-githooks: true

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

liuxuezhao committed:
DAOS-13564 container: add daos_cont_set_ro() API definition

1. add daos_cont_set_ro() API definition
2. add DAOS_PROP_CO_READONLY container status bit
3. rename daos_cont_status_clear() to daos_cont_set_healthy()
   add daos_cont_status_clear alias to keep back-compatibility
4. define DAOS_PROP_CSF_HEALTHY/DAOS_PROP_CSF_READONLY internal flag bit
   cleanup DAOS_PROP_CO_CLEAR as no usage in code.

Required-githooks: true

Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liuxuezhao committed:
DAOS-13564 container: a few update of API comment

Required-githooks: true
Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
